services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusDataTest.kt

```diff
@@ -24,293 +32,790 @@ import org.junit.jupiter.params.provider.NullSource
+                    ),
                 )
 
-            @ParameterizedTest
-            @EmptySource
-            @NullSource
-            fun `returns a change when areas were empty but new ones are added`(
-                areas: List<JohtoselvitysHakemusalue>?
+            @ParameterizedTest(name = "{displayName} {1}")
+            @MethodSource("johtoselvitysCases")
+            fun `returns changes when johtoselvityshakemus fields have changes`(
+                updated: JohtoselvityshakemusData,
+                name: String,
             ) {
-                val base = base.copy(areas = areas)
+                val changes = base.listChanges(updated)
+
+                assertThat(changes).containsExactly(name)
+            }
+        }
+
+        @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+        @Nested
+        inner class Kaivuilmoitus {
+
+            val base: KaivuilmoitusData = HakemusFactory.createKaivuilmoitusData()
+
+            @Test
+            fun `throws exception when classes don't match`() {
+                val updated = HakemusFactory.createJohtoselvityshakemusData()
 
-                val result = base.listChanges(dataWithTwoAreas)
+                val failure = assertFailure { base.listChanges(updated) }
 
-                assertThat(result).containsExactly("areas[0]", "areas[1]")
+                failure.all {
+                    hasClass(IncompatibleHakemusDataException::class)
+                    messageContains("Incompatible hakemus data when retrieving changes")
+                    messageContains("firstClass=KaivuilmoitusData")
+                    messageContains("secondClass=JohtoselvityshakemusData")
+                }
             }
 
-            @ParameterizedTest
-            @EmptySource
-            @NullSource
-            fun `returns a change when there were areas but now they are empty`(
-                areas: List<JohtoselvitysHakemusalue>?
-            ) {
-                val updated = base.copy(areas = areas)
+            @Test
+            fun `returns empty when there are no changes`() {
+                val updated = HakemusFactory.createKaivuilmoitusData()
 
-                val result = dataWithTwoAreas.listChanges(updated)
+                val result = base.listChanges(updated)
 
-                assertThat(result).containsExactly("areas[0]", "areas[1]")
+                assertThat(result).isEmpty()
             }
 
             @Test
-            fun `returns several changes when an area is removed from the middle`() {
-                val base =
+            fun `returns all changes when there are several`() {
+                val updated =
+                    base.copy(name = "Other name", startTime = base.startTime!!.plusDays(3))
+
+                val result = base.listChanges(updated)
+
+                assertThat(result).containsExactly("name", "startTime")
+            }
+
+            private fun commonFieldCases(): List<Arguments> =
```

Comment:
Is there a reason to repeat these tests? Both the code and the data interfaces are shared between the two types of applications.